eppo_core/src/events/event_dispatcher.rs

```diff
@@ -122,28 +108,36 @@ impl<T: EventQueue + Clone> EventDispatcher<T> {
                 }
             }
 
-            // Send events.
+            let event_delivery = event_delivery.clone();
+            let batch = event_queue.next_batch(QueuedEventStatus::Pending);
+            let queue = event_queue.clone();
+            // Send events
             tokio::spawn({
-                let event_delivery = event_delivery.clone();
-                let batch = event_queue.next_batch(QueuedEventStatus::Pending);
                 async move {
                     // Spawning a new task, so the main task can continue batching events and respond to
                     // commands. At this point, batch_queue is guaranteed to have at least one event.
                     let events = batch
-                        .into_iter()
-                        .map(|queued_event| queued_event.event)
+                        .iter()
+                        .map(|queued_event| queued_event.clone().event)
                         .collect();
                     let result = event_delivery.deliver(events).await;
                     match result {
                         Ok(response) => {
-                            if !response.failed_events.is_empty() {
-                                // TODO: Update status, increment attempts and re-enqueue events for retry
-                                warn!("Failed to deliver {} events", response.failed_events.len());
+                            let failed_event_uuids = response.failed_events;
+                            if !failed_event_uuids.is_empty() {
+                                warn!("Failed to deliver {} events", failed_event_uuids.len());
+                                let failed_events = batch
+                                    .into_iter()
+                                    .filter(|queued_event| {
+                                        failed_event_uuids.contains(&queued_event.event.uuid)
```

Comment:
Are you trying to prevent duplicates if the deleted event is still queued? If so I think that's too heavy handed. What do you think the system behavior looks like without this line.

Reply:
no, that's not the intention. the goal here is that `response.failed_events` may have a subset of the uuids of events from `batch`, as in, some events in a batch might fail to ingest, so we need to find them and mark as failed for retry. does that make sense?